wdl-format/src/token/post.rs

```diff
@@ -0,0 +1,256 @@
+//! Postprocessed tokens.
+//!
+//! Generally speaking, unless you are working with the internals of code
+//! formatting, you're not going to be working with these.
+
+use wdl_ast::SyntaxKind;
+
+use crate::Comment;
+use crate::LineSpacingPolicy;
+use crate::NEWLINE;
+use crate::PreToken;
+use crate::SPACE;
+use crate::Token;
+use crate::TokenStream;
+use crate::Trivia;
+use crate::config::Indent;
+
+/// A postprocessed token.
+#[derive(Eq, PartialEq)]
+pub enum PostToken {
+    /// A space.
+    Space,
+
+    /// A newline.
+    Newline,
+
+    /// One indentation.
+    Indent,
+
+    /// A string literal.
+    Literal(String),
+}
+
+impl std::fmt::Debug for PostToken {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Space => write!(f, "<SPACE>"),
+            Self::Newline => write!(f, "<NEWLINE>"),
+            Self::Indent => write!(f, "<INDENT>"),
+            Self::Literal(value) => write!(f, "<LITERAL> {value}"),
+        }
+    }
+}
+
+impl PostToken {
+    /// Returns a displayable version of the token.
+    pub fn display<'a>(&'a self, config: &'a crate::Config) -> Box<dyn std::fmt::Display + 'a> {
```

Comment:
I was struggling with the borrow checker. I'm not sure how to write this without an allocation, but I've updated it to call `to_string()` instead of `Box::new()`. Not sure if that's any better 🤷

Reply:
The trick is usually to define a local type to the function that implements the implements the returned interface and captures what it needs to operate.

For example, we could do it this way:

```rust
/// Returns a displayable version of the token.
fn display<'a>(&'a self, config: &'a crate::Config) -> impl Display + 'a {
    struct Display<'a> {
        token: &'a PostToken,
        config: &'a crate::Config,
    }

    impl std::fmt::Display for Display<'_> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            match self.token {
                PostToken::Space => write!(f, "{SPACE}"),
                PostToken::Newline => write!(f, "{NEWLINE}"),
                PostToken::Indent => {
                    let (c, n) = match self.config.indent() {
                        Indent::Spaces(n) => (' ', n),
                        Indent::Tabs(n) => ('\t', n),
                    };

                    for _ in 0..n.get() {
                        write!(f, "{c}")?;
                    }

                    Ok(())
                }
                PostToken::Literal(value) => write!(f, "{value}"),
            }
        }
    }

    Display {
        token: self,
        config,
    }
}
```

Here there's no internal allocations to do the formatting. We're simply wrapping the parameters with a type that implements `Display`; effectively this is like a closure over the token and the config so that we can use them from within a standard `Display` implementation.

Reply:
So I think I'm following along here. I needed to make some lifetime annotations elsewhere to make your edits work. Pretty sure I did that right, but this aspect of Rust still gives me some trouble. LMK if the edits look good!